### ISSUE TITLE
Icecream hangs while building, daemons show "unexpected eof" error messages

### ISSUE THREAD

**Issue by marlonsmith10** [icecc/icecream#294]
Hey guys,

I've compiled and installed icecc 1.1 and I can build my project with it successfully.  Work is distributed to all the nodes and compilation works ok.  The problem I'm getting is that icecc will often hang for a few minutes with no activity.  Every time this happens, one or more of the nodes shows the following:

[27144] 18:51:29: saw eof without complete msg! 1
[27144] 18:51:29: unexpected EOF while reading preprocessed file
[27144] 18:51:29: </parent, waiting: 1ms>
[27134] 18:51:29: timeout while reading preprocessed file

The result of this is that compilation takes longer with icecc than it would if I just used my local computer :/

Network connections are wired and fast.  I got the same result with a wireless network instead.  Any ideas what could cause this?

Thanks!

**Comment by marlonsmith10** [comment]
Update:  I also see this message on the daemon running on the local computer (the same computer that is running the scheduler):

saw eof without complete msg! 1

**Comment by HenryMiller1** [comment]
I've seen that error once in a while, but I've never been able to figure out why. It happens rarely for me though, and thus doesn't impact build performance for me.

I'd love to see a solution, but someone needs to figure out why this happens before we can solve it.

**Comment by marlonsmith10** [comment]
I am happy to help debug if I can.  I'll see if I can get anything in the log files.

**Comment by marlonsmith10** [comment]
I have figured out how to prevent the daemon from hanging on my machines and killing my performance.  I don't understand the icecc source enough to say if this is the best fix, but perhaps this can help the devs solve the problem.

In the source code for release 1.1, in the file workit.cpp, I added a line below line 458:

`} else if (client->at_eof()) {
                log_error() << "unexpected EOF while reading preprocessed file" << endl;
                return_value = EXIT_IO_ERROR;
                client_fd = -1;
                kill(pid, SIGTERM);
                delete fcmsg;
                fcmsg = 0;
                return EXIT_DISTCC_FAILED;    // <- This line added by me
}`

Now that I look at my addition, maybe I should have returned EXIT_IO_ERROR instead.  Either way, I suspect that after this section of code is reached, it takes a while for work_it() to return, during which time the daemon will not take on additional tasks.  Eventually all of the machines on my cluster got into this state, and the whole thing came to a grinding halt for minutes at a time.  Once I added that line, performance increased significantly, and icecc now helps my compilation time instead of hindering it.
